### PR TITLE
Bug 1821464 Add UI test to NotificationWorker

### PR DESCRIPTION
**Pull request by jhugman**
Fixes [EXP-3184](https://mozilla-hub.atlassian.net/browse/EXP-3184).
Fixes [Bug 1821464](https://bugzilla.mozilla.org/show_bug.cgi?id=1821464).

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1821464

### CODE REVIEW

**Comment by jhugman** [comment]
Depends on #1159 landing.

File: fenix/app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt [hunk 1]
```diff
@@ -159,8 +160,13 @@ class MessageNotificationWorker(
             val instanceWorkManager = WorkManager.getInstance(context)
             instanceWorkManager.enqueueUniquePeriodicWork(
                 MESSAGE_WORK_NAME,
-                // We want to keep any existing scheduled work.
-                ExistingPeriodicWorkPolicy.KEEP,
+                // We want to keep any existing scheduled work, unless
+                // when we're under test.
+                if (messaging.isUnderTest()) {
```
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt:165]: I think ideally we would not have logic for testing purposes. Wonder if we modify HomeActivityIntentTestRule and pass in the ExistingPeriodicWorkPolicy to setMessageNotificationWorker

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingNotificationTest.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import org.junit.Test
+import org.mozilla.experiments.nimbus.HardcodedNimbusFeatures
+import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.helpers.HomeActivityIntentTestRule
+import org.mozilla.fenix.helpers.TestHelper
+import org.mozilla.fenix.nimbus.FxNimbus
+import org.mozilla.fenix.ui.robots.notificationShade
+
+/**
+ * A UI test for testing that surfaces work and remain working.
```
Comment (t-p-white) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingNotificationTest.kt:25]:
```suggestion
 * UI test for testing the Messaging Notification surface.
```

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingNotificationTest.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+ */
+class NimbusMessagingNotificationTest {
+    private lateinit var mDevice: UiDevice
+
+    private lateinit var context: Context
+    private lateinit var hardcodedNimbus: HardcodedNimbusFeatures
+
+    private val storage
```
Comment (t-p-white) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingNotificationTest.kt:33]: not used?

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingNotificationTest.kt [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                      "trigger": [
+                        "ALWAYS"
+                      ]
+                    }
+                  }
+                }
+                """.trimIndent(),
+            ),
+        )
+        activityTestRule.finishActivity()
```
Comment (t-p-white) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessagingNotificationTest.kt:78]: some comments around why we need to finish the activity and launch the activity may be useful

**Review by t-p-white — APPROVED** [review]
Nice work. Works as expected 👍 it would be nice if we could move the test logic out, but I'll leave that decision to you 🙂

**Pull request merged**